flowmachine/flowmachine/features/location/labelled_spatial_aggregate.py

```diff
@@ -0,0 +1,117 @@
+
+    >>> locations = locate_subscribers(
+    ...     "2016-01-01",
+    ...     "2016-01-02",
+    ...     spatial_unit=make_spatial_unit("admin", level=3),
+    ...     method="most-common",
+    ... )
+    >>> metric = SubscriberHandsetCharacteristic(
+    ...     "2016-01-01", "2016-01-02", characteristic="hnd_type"
+    ... )
+    >>> labelled = LabelledSpatialAggregate(locations=locations, subscriber_labels=metric)
+    >>> labelled.get_dataframe()
+               pcod label_value  value
+     0  524 3 08 44     Feature     36
+     1  524 3 08 44       Smart     28
+     2  524 4 12 62     Feature     44
+     3  524 4 12 62       Smart     19'
+
+    """
+
+    def __init__(
+        self,
+        locations: Query,
+        subscriber_labels: Query,
+        label_columns: List[str] = ("value",),
+    ):
+
+        if any(
+            label_column not in subscriber_labels.column_names
+            for label_column in label_columns
+        ):
+            raise ValueError(
+                f"One of {label_columns} not a column of {subscriber_labels}"
+            )
+        if "subscriber" not in locations.column_names:
+            raise ValueError(f"Locations query must have a subscriber column")
+        if "spatial_unit" not in dir(locations):
+            raise ValueError(f"Locations must have a spatial_unit attribute")
+        if "subscriber" in label_columns:
+            raise ValueError(f"'subscriber' cannot be a label")
+
+        self.locations = locations
+        self.subscriber_labels = subscriber_labels
+        self.label_columns = list(label_columns)
+        self.spatial_unit = locations.spatial_unit
+        self.label_columns = label_columns
+        self.out_label_columns = [
+            f"label_{label_col}" for label_col in self.label_columns
+        ]
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return (
+            list(self.spatial_unit.location_id_columns)
+            + [f"label_{label}" for label in self.label_columns]
+            + ["value"]
+        )
+
+    @property
+    def out_label_columns_as_string_list(self):
+        """
+        Returns the label column heading as a single string
+        """
+        return ",".join(self.out_label_columns)
+
+    def _make_query(self):
+
+        aggregate_cols = ",".join(
+            f"agg.{agg_col}" for agg_col in self.spatial_unit.location_id_columns
+        )
+        label_select = ",".join(
+            f"labels.{label_col} AS {out_label_col}"
+            for label_col, out_label_col in zip(
+                self.label_columns, self.out_label_columns
+            )
+        )
+        label_group = ",".join(
+            f"labels.{label_col}" for label_col in self.label_columns
+        )
+
+        sql = f"""
+            SELECT
+                {aggregate_cols}, {label_select}, count(*) AS value
+            FROM 
+                ({self.locations.get_query()}) AS agg
+            JOIN
```

Comment:
Might want to use a left join here, I think - which would lead to an additional group (label=NULL) for subscribers that appear in the location query but not the label query.